SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorCloudLauncherSettings.cpp

```diff
@@ -0,0 +1,160 @@
+	FString SpatialOSFile = GetDefault<USpatialGDKEditorSettings>()->GetSpatialOSDirectory().Append(TEXT("/spatialos.json"));
+
+	if (!FFileHelper::LoadFileToString(FileContents, *SpatialOSFile))
+	{
+		return TEXT("");
+	}
+
+	TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(FileContents);
+	TSharedPtr<FJsonObject> JsonObject;
+
+	if (FJsonSerializer::Deserialize(Reader, JsonObject))
+	{
+		return JsonObject->GetStringField("name");
+	}
+
+	return TEXT("");
+}
+
+void USpatialGDKEditorCloudLauncherSettings::ValidateAssemblyName()
+{
+	const FRegexPattern AssemblyPattern(TEXT("^[a-zA-Z0-9_.-]{5,64}$"));
+	FRegexMatcher RegMatcher(AssemblyPattern, AssemblyName);
+
+	bAssemblyNameIsValid = RegMatcher.FindNext();
+}
+
+void USpatialGDKEditorCloudLauncherSettings::ValidateProjectName()
+{
+	const FRegexPattern ProjectPattern(TEXT("^[a-z0-9_]{3,32}$"));
+	FRegexMatcher RegMatcher(ProjectPattern, ProjectName);
+
+	bProjectNameIsValid = RegMatcher.FindNext();
+}
+
+void USpatialGDKEditorCloudLauncherSettings::ValidateDeploymentName()
+{
+	const FRegexPattern DeploymentPattern(TEXT("^[a-z0-9_]{2,32}$"));
+	FRegexMatcher RegMatcher(DeploymentPattern, PrimaryDeploymentName);
+
+	bPrimaryDeploymentNameIsValid = RegMatcher.FindNext();
+}
+
+void USpatialGDKEditorCloudLauncherSettings::SetPrimaryDeploymentName(const FString & Name)
+{
+	PrimaryDeploymentName = Name;
+}
+
+void USpatialGDKEditorCloudLauncherSettings::SetAssemblyName(const FString & Name)
+{
+	AssemblyName = Name;
+}
+
+void USpatialGDKEditorCloudLauncherSettings::SetProjectName(const FString & Name)
+{
+	ProjectName = Name;
+}
+
+void USpatialGDKEditorCloudLauncherSettings::SetPrimaryLaunchConfigPath(const FString & Path)
+{
+	PrimaryLaunchConfigPath.FilePath = FPaths::ConvertRelativePathToFull(Path);
+}
+
+void USpatialGDKEditorCloudLauncherSettings::SetSnapshotPath(const FString & Path)
+{
+	SnapshotPath.FilePath = FPaths::ConvertRelativePathToFull(Path);
+}
+
+void USpatialGDKEditorCloudLauncherSettings::SetSimulatedPlayersEnabledState(bool IsEnabled)
+{
+	SimulatedPlayersIsEnabled = IsEnabled;
+}
+
+void USpatialGDKEditorCloudLauncherSettings::SetSimulatedPlayerDeploymentName(const FString & Name)
+{
+	SimulatedPlayerDeploymentName = Name;
+}
+
+void USpatialGDKEditorCloudLauncherSettings::SetSimulatedPlayerLaunchConfigPath(const FString & Path)
+{
+	SimulatedPlayerLaunchConfigPath.FilePath = FPaths::ConvertRelativePathToFull(Path);
+}
+
+void USpatialGDKEditorCloudLauncherSettings::SetNumberOfSimulatedPlayers(uint32 Number)
+{
+	NumberOfSimulatedPlayers = Number;
+}
+
+bool USpatialGDKEditorCloudLauncherSettings::IsDeploymentConfigurationValidSinceLastCheck() const
+{
+	return bProjectNameIsValid &&
+		bAssemblyNameIsValid &&
+		bPrimaryDeploymentNameIsValid &&
+		!PrimaryLaunchConfigPath.FilePath.IsEmpty() &&
+		!SnapshotPath.FilePath.IsEmpty();
+}
+
+bool USpatialGDKEditorCloudLauncherSettings::IsDeploymentConfigurationValidWithCheck()
+{
+	ValidateAssemblyName();
```

Comment:
Can't we have these functions return the result of the validation and pass them on to `IsDeploymentConfigurationValidSinceLastCheck`? that way we can early out and get rid of the member variables. At that point we can also make the validation functions const. What do you think?

Reply:
The reason for the member variables is that we can use the function `IsDeploymentConfigurationValidSinceLastCheck()`, which only checks the member variables rather than validate the regex every time. The problem with not having the member variables is that if we decide to use that function to decide whether or not a specific button is enabled (e.g. 'Launch Cloud Deployment' button is enabled only if all the configurations are valid) then the function is called quite often (every frame? - not sure), which means that the regex is checked a lot. It's just something that didn't sit right with me when I was writing it 😄. Alternatively, it could just be one function that validates everything when called.